Polymorphic Parameters

voodoos opened the pull request:
This upstreaming PR adds the possibility to have function parameters with polymorphic types in the language.

This work was [originally done by @lpw25](https://github.com/ocaml-flambda/ocaml-jst/pull/14) in Jane Street's fork of the compiler with extensions. Jane Street has asked for help from Tarides to upstream these changes.

A notable part of the implementation is the introduction of a new invariant: 

> The argument's type in a `Types.Tarrow` node must be a `Tpoly` node. 
> Monomorphic arguments are represented by a `Tpoly` node with an empty list of names.

Here is the original description of the feature written by @lpw25:

## Polymorphic parameters

The *polymorphic parameters* extension allows you to have function
parameters with polymorphic types. For example, a function could
have a type like:
```ocaml
val f : ('a. 'a -> 'a list) -> int list
```

As a practical example, let's consider the creation functions from
`ppx_typed_fields`. Given a record definition:
```ocaml
type t =
  { a : string
  ; b : int
  }
```
`ppx_typed_fields` gives you a type representing the fields indexed by
their type:
```ocaml
type 'a field =
  | A : string field
  | B : int field
```
A useful function to provide for these types is one that can create a
`t` if given a function that returns a value for each of the fields. You
might try to write that function like so:
```ocaml
let create f =
  { a = f A; b = f B }
```
but the compiler will complain about this:
```ocaml
Line 3, characters 21-22:
3 |     { a = f A; b = f B };;
                         ^
Error: This expression has type int field
       but an expression was expected of type string field
       Type int is not compatible with type string
```
The issue is that you need to apply `f` to both a `string field` and
an `int field`: you need it to be *polymorphic*.

There are existing ways to work around this issue -- polymorphic
record fields, first-class modules and polymorphic methods -- but they
all require defining a fresh type to contain `f`. These work-arounds
require additional code at each call site to wrap `f` in the associated
type.

With polymorphic parameters, you can simply annotate `f` as
polymorphic:
```ocaml
let create (f : 'a. 'a field -> 'a) =
  { a = f A; b = f B }
```
which gives `create` the following type:
```ocaml
val create : ('a. 'a field -> 'a) -> t
```

It can be called on a suitable function directly:
```ocaml
let forty_two (type a) : a field -> a = function
  | A -> "forty two"
  | B -> 42

let r = create forty_two
```

## Limitations

All polymorphic parameter require an annotation. Without an annotation
OCaml will assume that the parameter is monomorphic.

Rather than annotating `f` directly, you can also annotate `create` as
a whole:
```ocaml
let create : ('a. 'a field -> 'a) -> t =
  fun f -> { a = f A; b = f B }
```
or rely on the expected type propagated from a function application:
```ocaml
let with_creator (c : ('a. 'a field -> 'a) -> t) =
  c forty_two

let r = with_creator (fun f -> { a = f A; b = f B })
```

Type variables in OCaml always stand for monomorphic types. You cannot
instantiate a type like `'a -> 'b` to `('a. 'a field -> 'a) -> t`. In
practice, this means that functions like `apply`:
```ocaml
let apply f x = f x
```
cannot be applied to functions with polymorphic parameters:
```ocaml
Line 1, characters 6-12:
1 | apply create forty_two;;
          ^^^^^^
Error: This expression has type ('a. 'a field -> 'a) -> t
       but an expression was expected of type 'b -> 'c
       The universal variable 'a would escape its scope
```

Similarly, applying a function with polymorphic parameters requires
knowing the type of the function. Without type information, OCaml will
assume that the function has type `'a -> 'b`. The rules for knowing
the type of the function are the same as those for record field
disambiguation. For example,
```ocaml
let with_create_mono c =
  c forty_two
```
will get the type:
```ocaml
val with_create_mono : (('a field -> 'a) -> 'b) -> 'b
```
because OCaml has assumed that the parameter of `c` is monomorphic,
which prevents it from being applied to `create`:
```ocaml
Line 1, characters 17-23:
1 | with_create_mono create;;
                     ^^^^^^
Error: This expression has type ('a. 'a field -> 'a) -> t
       but an expression was expected of type ('b field -> 'b) -> 'c
       The universal variable 'a would escape its scope
```

Note that these limitations only apply to parameters with genuinely
polymorphic parameters. Given a type like:
```ocaml
type 'a t = int
```
the type `('a. 'a t) -> string` is completely equivalent to `int ->
string`.

#### Remaining tasks
- [x] Fix the remaining regression
- [ ] Restore authorship of commits

cc @OlivierNicole @goldfirere @lpw25

lpw25 committed:
Add tests for polymorphic parameters

voodoos committed:
Introduce new error "opt params cannot be polymorphic"

voodoos committed:
Wrap and generalize filter_arrow_mono if principal

voodoos committed:
Refactor Pexp_function case and handle poly

voodoos committed:
Add one generalization level in type_class_arg_pattern

voodoos commented on testsuite/tests/typing-misc/let_rec_approx.ml:
```diff
@@ -33,6 +33,9 @@ Error: This pattern matches values of type "string"
 and opt_ok_g ?(foo : M.t option) ?(bar : M.t = M.A) () = opt_ok_f ()
 [%%expect{|
-val opt_ok_f : unit -> 'a = <fun>
-val opt_ok_g : ?foo:M.t -> ?bar:M.t -> unit -> 'a = <fun>
+Line 2, characters 35-44:
+2 | and opt_ok_g ?(foo : M.t option) ?(bar : M.t = M.A) () = opt_ok_f ()
+                                       ^^^^^^^^^
+Error: This pattern matches values of type "M.t"
+       but a pattern was expected which matches values of type "'a option"
 |}]
```
Hi @stedolan,

You introduced that test in #12428 to illustrate a regression after the improvement to type-based disambiguation you made in #12315. None of these patches are included in flambda backend and all the related tests fail there. 

This PR includes a refactoring of `type_approx` in 7424f438630451a019f8d0e2c7cb0818f8e33630 and I was able to re-apply your improvement (155690a72f1d0f4d671691798a2967f73b6c1d69), but so far I failed to do the same with the follow-up fix (84e4d474de4acea497dabc1e8777386d6506fa03). Would you, by any chance, have some time to look at it ?

stedolan replied:
I'll have a look at this, but I think it'll be next week before I find time.

voodoos replied:
Thank you, meanwhile I will follow Leo's suggestion in https://github.com/ocaml/ocaml/pull/13806#discussion_r1958427530 and see if it leads me to a solution.

voodoos replied:
This is sorted out in [7fecb8a](https://github.com/ocaml/ocaml/pull/13806/commits/7fecb8aeb93cd64d3baa3c075235f66c5ad2e2a8)

samsa1 commented:
From what I remember from a discussion with @goldfirere we concluded that we didn't want to have the invariant :

> The argument's type in a `Types.Tarrow` node must be a `Tpoly` node.
> Monomorphic arguments are represented by a `Tpoly` node with an empty list of names.

If I remember correctly, it was done this way on flambda backend because they wanted to minimize the diff with the upstream compiler. However we don't have such a constraint here and I feel that encoding invariants directly into the AST is better than having the risk of introducing a bug because one forgot a call to `Ctype.newmono`.

However it might significantly increase the diff in `Ctype`, have you tried to look at the impact of the following change ?

```ocaml
| Tarrow of arg_label * (type_expr * type_expr list) * type_expr * commutable
```

Maybe the diff could be reduced by introduction a new type for `Tpoly`s (a record with the fields `type_expr` and `type_expr list`) leading to a factorization between the code for `Tpoly`s and the modified `Tarrow`.

Octachron commented:
Personally, the diff size of mechanical changes does not matter much to me when reviewing. If the difference increase too much, I would rather propose to first introduce an inline record in a separate commit, then add the new field.

samsa1 commented on testsuite/tests/typing-objects/Exemples.ml:
```diff
@@ -547,10 +547,7 @@ Error: Type
-       is not a subtype of
-         "int_comparable2" =
-           "< cmp : int_comparable2 -> int; set_x : int -> unit; x : int >"
+       Type "int_comparable" is not a subtype of "int_comparable2"
```
Do you have an idea of why we lost the description of the types here ?

lpw25 replied:
Is the terrible code that uses `Marshal` to tell if something is big still there? That used to cause this kind of change whenever someone changed `types.ml`.

Octachron replied:
The terrible code is still here, but only for module inclusion errors, and the check is disabled by default in expect tests.

Octachron replied:
A somewhat simpler reproduction case:
```ocaml
type u0 = int * int -> unit
type u = u0 -> unit
type v0 = float * float -> unit
type v = v0 -> unit
let f = (g: u -> unit :> v -> unit)
```
The issue is that `Ctype.subtype` creates a subtype trace that contains function parameters prefixed with an empty `Tpoly` while the expansion function `Ctype.full_expand` for the error never expand such types.

The fix is probably to remove empty matching `Tpoly`  before adding them to the subtyping trace  in `subtype_rec`.

Octachron replied:
Here is a -1 lines fix: https://github.com/Octachron/ocaml/commit/3084345fb0ec98f5a35b1ae90331c1af4b6ebc76 .

voodoos replied:
Thanks, I added the test and the fix to the PR.

gasche commented on typing/typecore.ml:
```diff
@@ -3284,11 +3269,15 @@ let type_approx_fun_one_param env label spato ty_expected ~first ~in_function =
+        type_pattern_approx env spat ty_arg
+      else
+        type_pattern_approx env spat (Btype.tpoly_get_mono ty_arg)
+  in
```
This smells weird to me. What are these branches supposed to do? Could we do the same by checking for the case `Ptyp_poly(ty, [])` in `approx_type`?

lpw25 replied:
I agree that this looks off. I think the old code from my PR was trying to only look at type annotations when they were polymorphic so as not to change the existing behaviour for other annotations. Now that #12315 has made it look at all annotations, we probably don't need to distinguish the polymorphic case so much.

voodoos replied:
I indeed added the second branch to fix the tests introduced in #12315.
At first I simply tried to remove the `has_poly` condition but it doesn't work without unwrapping the tpoly.

lpw25 replied:
Looking at what `type_function` does, you should probably be doing this logic here: https://github.com/ocaml/ocaml/blob/95c5177ef97bdb6524273297ca27cf853febdac0/typing/typecore.ml#L5202

gasche commented:
Looking at the commits one-by-one, in many cases I feel a bit lost as I don't know what the purpose of the commit is. (Bonus point for the commit message "adapt to the new Tpoly invariant", with no explanation of what this new invariant actually is.)

For example https://github.com/ocaml/ocaml/pull/13806/commits/01adae10d96ad6311d20b6fc80a936dd835e00dc , I can see that this commit is duplicating a principality check from the previous commit, but why? There is no mention of why the code is needed here, no change to the testsuite, nothing that can let me guess why this change was deemed necessary.

gasche commented:
If the PR is intended to be reviewed commit-by-commit, I think that it would be nice if the commits would come with a commit message that explains the change, maybe give an example if that helps. (Sometimes it can be helpful to run the testsuite and include the testsuite changes in the commit itself, as a way to understand the impact.)

voodoos commented:
> From what I remember from a discussion with @goldfirere we concluded that we didn't want to have the invariant
> ```ocaml
> | Tarrow of arg_label * (type_expr * type_expr list) * type_expr * commutable
> ```

I think that makes  a lot of sense, I will give your suggestion a try.

@gasche thanks for the quick-look, I will rework the commit history after having removed the "invariant". It is sometimes hard to de-compile a single commit into meaningful steps, and I obviously failed here.

johnyob commented on typing/ctype.mli:
```diff
@@ -283,9 +284,26 @@ val unify_gadt:
+           argument type be a [Tpoly] node is not enforced. Raises
+           [Filter_arrow_failed] instead of [Unify].  *)
+val filter_mono: type_expr -> type_expr
+        (* A special case of unification (with Tpoly('a, [])). Can
+           only be called on [Tpoly] nodes. Raises [Filter_mono_failed]
+           instead of [Unify] *)
```
The doc string here doesn't match the behaviour of `filter_mono`, namely in the case of `filter_mono 'b`, we do not unify `'b` with `Tpoly('a, [])`

Octachron replied:
I agree that it feels that the function name doesn't match the function effect. In fact, I would propose to move the function to `Btype` with the other `Tpoly` as `tpoly_get_mono_opt`  and update its type to `type_expr -> type_expr opt`.

johnyob commented on typing/btype.mli:
```diff
@@ -120,6 +121,13 @@ val proxy: type_expr -> type_expr
+(**** Utilities for poly types ****)
+
+(* These three functions can only be called on [Tpoly] nodes. *)
+val tpoly_is_mono : type_expr -> bool
```
Wouldn't it be more consistent with `is_poly_Tpoly` for this to be `is_mono_Tpoly`?

Octachron replied:
The `tpoly_*` functions share the precondition that their argument must be a `Tpoly`, which is not the case for `is_poly_Tpoly`. Thus I think the current name is fine.

lpw25 commented:
> If I remember correctly, it was done this way on flambda backend because they wanted to minimize the diff with the upstream compiler.

It is actually done that way because that is how the current implementation handles all such cases -- that is why `Tpoly` already exists and is used for e.g. method fields. If someone wants to investigate getting rid of `Tpoly` from the type-checker and using a separate type for polymorphic types, then I suggest they do so in a separate PR.

lpw25 commented:
> I think that makes a lot of sense, I will give your suggestion a try.

I would hold off on trying to make that change. I think it will require duplicating a lot of code needlessly between polymorphic parameters and polymorphic methods. To me it just seems completely orthogonal to the current PR.

Octachron commented:
Maybe an intermediary solution would to be make the type of `Tarrow` arguments an abstract `tpoly`, since the code looks like it is already using constructors and destructors functions for those arguments.

samsa1 commented on typing/typecore.ml:
```diff
@@ -5050,19 +5184,33 @@ and split_function_ty env ty_expected ~arg_label ~first ~in_function =
+      let ty, vars = tpoly_get_poly ty_arg in
+      if vars = [] then ty
+      else begin
+        with_level ~level:generic_level
+          (fun () -> snd (instance_poly ~keep_names:true ~fixed:false vars ty))
```
I found a bug linked to principality in this PR.

When testing
```ocaml
let foo3 (x : 'a. 'a -> 'a as 'b) (y : 'b) = y 3
```
we don't get a principality issue.

However, the following example raises one
```ocaml
let foo2 (x : a:int -> b:int -> int as 'b) (y : 'b) =
    y ~b:3 ~a:1
```

The lines above raise the level of the generic structure of the type `ty`, thus marking principal non-principal information. The bug most likely comes from those lines but could arrive from somewhere else.

samsa1 replied:
The example above is wrong due to an error in priority between `'a. _` and `_ as _` 

It seems that a few lines above we do a principality check so raising the level here might be correct. However we might have a problem if `ty = 'b. x:unit -> y:unit -> 'b t` in a non principal way and `'b t = unit`.
I haven't managed to create an example where such a case could happen because `lower_level` removes the `'b t`. Is there a way to be certain such a case cannot happen ?

garrigue commented:
I am probably not going to be able to look at this PR in detail for a few weeks.
Two remarks.
> There are existing ways to work around this issue -- polymorphic
> record fields, first-class modules and polymorphic methods -- but they
> all require defining a fresh type to contain f. These work-arounds
> require additional code at each call site to wrap f in the associated
> type.

The first sentence is incorrect for polymorphic methods: the is no need to define a new type, you just need to annotate with an object type:
```ocaml
let create (f : <m : 'a. 'a field -> 'a>) =
  { a = m#f A; b = m#f B }
```
However, you also need to write this type when you wrap the function, so the second sentence is right.

The other is that I am a bit dubious about the `with_creator` example.
IIRC, this kind of propagation is not principal.
The other examples seem OK, since there is either a direct annotation, or it is propagated syntactically.
This said, an example similar to `with_creator` is accepted for GADTs, so we have to decide what is OK.
(To be more precise, this is principal in the type system restricted to minimal derivations)
```ocaml
let f (type a) (x : a field) = let g (h : a field -> a) = h x in g (function A -> true | B -> 1);;
```

samsa1 commented on parsing/parser.mly:
```diff
@@ -2391,6 +2395,11 @@ label_let_pattern:
+          cty = mktyp (vars = typevar_list DOT ty = core_type { Ptyp_poly(vars, ty) })
+      { let lab, pat = x in
+        lab,
+        mkpat ~loc:$sloc (Ppat_constraint (pat, cty)) }
```
Wouldn't it be nicer to introduce a new parsing rule here ?
```ocaml
maybe_core_type:
  | cty = core_type { cty }
  | poly_case
```

It could be used here and in the `let_pattern` case bellow

Octachron replied:
This sounds like  'poly_type' which is already defined.

Octachron replied:
This rule can be unified with the one above by replacing `core_type|poly_type` by `possibly_poly_type`.

Octachron replied:
Sorry @OlivierNicole , it seems that this comment was hidden in the resolved conversation. Also the menhir version used for the bootstrap should be the same or a more recent one than the one used previously (20240715).

samsa1 commented on parsing/parser.mly:
```diff
@@ -3716,6 +3736,15 @@ function_type:
+  | mktyp(
+    LPAREN vars = typevar_list DOT ty = core_type RPAREN
+      { Ptyp_poly(vars, ty) }
+    )
```
Could maybe benefit from a refactoring of the parsing rule for `Ptyp_poly`

voodoos committed:
Promote polymorphic parameters tests failures

voodoos committed:
Refactor: more descriptive return type for [split_function], renaming and comment work

voodoos committed:
Handle polymorphic parameters in type_apply_arg

voodoos commented:
Thanks for your suggestion of extracting the refactor of type_approx, @samsa1, it did lead to a fix for the issue that was remaining. I also reverted the two commits adding levels without any observable changes, @lpw25 do you remember why you needed this, is it specific to ocaml-flambda ? (https://github.com/ocaml/ocaml/pull/13806/commits/ca118b6f6aa9aa8a37708bbc50c969360e8969b5 and https://github.com/ocaml/ocaml/pull/13806/commits/0f32e5ce3432c079b59273bbecdd626de95f38aa) 

I am in the process of rebasing on trunk, which is fairly non-trivial, but I think the PR will be ready for a round of review then.

voodoos committed:
Add changes entry for #15733

lpw25 commented:
I'm not sure why they are in this patch -- probably I just fixed them as I was going past -- but they both look like correct bug fixes to me. Probably if you take them out you can get something to complain about principality even though it shouldn't.

voodoos commented:
> Probably if you take them out you can get something to complain about principality even though it shouldn't.

That's not the case, they are reverted right now and that has no impact on the testsuite (maybe they do in flambda-backend ?).

lpw25 commented:
I didn’t mean something in the testsuite would fail. I meant that, with some effort, it should be possible to produce an example of misbehaviour that they fix. Feel free to take them out though, I doubt anyone will hit such cases in practice: barely anyone turns on principal mode anyway.

garrigue commented:
I had a (quick) look at the code, and I did not find anything untoward.
My only concern with respect to principality is how to properly define upward propagation through the expected type.
IIRC it was introduced for GADTs, and the main intent was to do a form of syntactic propagation of type annotations given on a function definition to its formal parameters and body (both for allowing type refinement in pattern-matching and avoiding ambivalence). Rather than going the syntactic way, which would have required a separate pass, we went for directly propagating the type. For GADTs, this did not matter much, as uses of GADTs involve pattern matching, which is rather verbose, so that one does not write a function pattern-matching on a GADT directly inside a function argument or a data structure. But first-class polymorphism is more light-weight, and one might be tempted to do that, as some examples in the testsuite suggest. It would be good to have a precise definition of which constructions propagate the expected type in a "principal" way (principal here is only related to the flag, because there is actually no way to make it really principal).
My impression is that we want to propagate the type inside a function definition, and also from a function to its argument (not so clear here, but needed for first class modules and optional argument discarding). But probably nothing else, because then it becomes much more difficult to define properly. We are maybe already doing too much.
We might also want to improve the theory, for instance using these new polymorphic arguments to track what is know deep inside their own type, but this would require further changes in the type system.

lpw25 commented:
> My impression is that we want to propagate the type inside a function definition, and also from a function to its argument (not so clear here, but needed for first class modules and optional argument discarding). But probably nothing else, because then it becomes much more difficult to define properly.

I think I'd probably want to include a few more things that this -- although perhaps you were already including some of these implicitly -- for instance:
- Right-hand side of `if`
- Right-hand side of `match`
- Body of a `try`
- Arguments of constructors
- Arguments of records

garrigue commented:
> I think I'd probably want to include a few more things that this -- although perhaps you were already including some of these implicitly -- for instance:

I would like to keep the list minimal: any formal description or proof about those will require going through all the supported cases.
This said, you are right that right-hand-side of match should be included, otherwise ambivalence resolution would not work in many cases, since this specifically the place where scope changes. For the other cases, I do not see strong motivation. A subtle case is arguments of constructors and records: one would expect to propagate at least as much as function applications; i.e. the type of their definition, but not necessarily propagation of parameters.
Also expectations might change with modular implicits; but I would like a clean theory for them.

lpw25 commented:
>  I do not see strong motivation

Consistency is important. If we are doing it for `match` then we should do it for `if` and `try` as they are just syntax for special forms of `match`.

Constructors and records have actually always had more aggressive propagation than function applications. I think changing that now would break a lot of code.

samsa1 commented on typing/typecore.ml:
```diff
@@ -217,6 +217,18 @@ let not_principal fmt =
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
 
+let error_of_filter_arrow_failure ~explanation ~first ty_fun
```
This function should also be used in `split_function_ty`.

samsa1 commented on typing/typecore.ml:
```diff
@@ -3179,6 +3191,21 @@ let check_recursive_class_bindings env ids exprs =
+   to the end of the overall function. The parser does not construct such
+   a location so we forge one for type errors.
+*)
+let loc_rest_of_function
```
Should also be called in `type_function` as this is supposed to be a refactoring.

samsa1 commented on typing/typecore.ml:
```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
-       ty
-    | Optional _, Some _ ->
-       type_option ty
+let type_approx_fun_one_param
```
Could be nice to add a comment explaining with `spato` is an optional pattern (ie : because we skip a match potentially containing a gadt)

samsa1 commented on typing/typecore.ml:
```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
         raise (Error (loc, env, Expr_type_clash (err, None, None)))
       end;
-      ty_coerce
+      ty_expected
```
This type here is formally wrong. It is not the expected type of the expression inside the constraint. However as this type is never used (see `ignore` at call site) it is not problematic. But this type still looks correct which could lead to potential issue in the future. I propose to replace `ty_expected` here by a trivially wrong type such as `Tuple []` to prevent people from using it.

Another approach would be to refactor this function into only one case : taking a `approx_type_with_expected`

samsa1 replied:
This type was used later on an causes a problem in the following example :
```ocaml
module M : sig
   type t = private int
   val x : t
end = struct
   type t = int
   let x = 0
end

let rec f () : M.t :> int = (M.x : M.t)
```

We need to either return the approximation `type_approx constrain` or `newvar ()`

samsa1 commented on typing/typecore.ml:
```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
+  end;
+  List.iter2
+    (fun (_, e) (_, ty) -> type_approx env e ty)
+    l labeled_tys
```
Seems to change slightly the set of codes accepted by the type checker because we now go inside tuples.

This is a change in a good direction but adding a test illustrating it could be nice

samsa1 commented on typing/typecore.ml:
```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
+    l labeled_tys
+
+and type_approx_function =
+  let rec loop env params c body ty_expected ~in_function ~first =
```
Maybe not adding a auxiliary recursive function in a already recursive function.
This would be nicer to just use `type_approx_function` for the recursive case

samsa1 commented on typing/typecore.ml:
```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
+                ~in_function ~first
+            in
+            type_approx env e ty_res
+        | Pfunction_cases ([], _, _) ->  ()
```
This case in not reachable from the parser. I could be nice to add a comment explaining that here.
(the implementation is indeed what should be done if this case was reachable)

Octachron commented on parsing/parser.mly:
```diff
@@ -2393,6 +2403,17 @@ let_pattern:
       { Ppat_constraint($1, $3) })
       { $1 }
+  | poly_pattern
+      { $1 }
```
This rule looks like it can be merged with the rule above as

```ocaml
mkpat (pattern COLON poly_type)
```

voodoos committed:
Add test with class argument

Octachron commented on typing/errortrace_report.ml:
```diff
@@ -117,10 +117,14 @@ let print_tag ppf s = Style.inline_code ppf ("`" ^ s)
-  match Types.get_desc (Ctype.expand_head env ty) with
-  | Tconstr (p, _, _) -> Path.same p Predef.path_unit
-  | _ -> false
+let is_unit_arg env ty =
```
Probably more `is_unit_param`

voodoos committed:
Additional tests with labels

Octachron commented on typing/typecore.ml:
```diff
@@ -4729,7 +4766,7 @@ and type_expect_
         | Tvar _ ->
             let exp = type_exp env sbody in
-            let exp = {exp with exp_type = newty (Tpoly (exp.exp_type, []))} in
+            let exp = {exp with exp_type = newmono exp.exp_type} in
```
This feels like a coincidental change since this type is not a function parameter type.

Octachron commented on typing/typecore.ml:
```diff
@@ -5216,22 +5264,22 @@ and type_function
       :: rest
     ->
-      let ty_arg, ty_res =
+      let ty_arg, ty_res, ty_arg_mono =
```
Note this is the reverse of the usual convention that `ty_arg_mono` is the result of `newmono ty_arg`.

Octachron commented on typing/typecore.ml:
```diff
@@ -5822,14 +5872,16 @@ and type_apply_arg env (lbl, arg) =
         unify_exp ~sexp:sarg env arg (type_option(newvar()));
       (lbl, Arg arg)
   | Arg (Known_arg { sarg; ty_arg; ty_arg0; wrapped_in_some }) ->
+      let ty_arg', _vars = tpoly_get_poly ty_arg in
```
This should be a `tpoly_get_mono` in this commit.

voodoos committed:
Handle polymorphic parameters in [type_approx and] [type_pat_aux]

voodoos committed:
Update parser and adapt to the new Tpoly invariant. Requires bootstrap.

The argument of a Tarrow must now be a [Tpoly] node.

voodoos committed:
bootstrap

lpw25 committed:
Add test for polymorphic variants in polymorphic parameters

lpw25 committed:
Take fixed instances of polys where appropriate

This fixes a bug in the interaction between polymorphic variants and polymorphic
parameters. The actual bug fix is just changing two falses to trues but I also
changed the instance_poly interface to make similar mistakes less likely to
happen.

samsa1 commented on typing/ctype.ml:
```diff
@@ -1160,7 +1162,7 @@ let rec copy ?partial ?keep_names copy_scope ty =
         None -> assert false
       | Some (free_univars, keep) ->
-          if TypeSet.is_empty (free_univars ty) then
+          if not (is_Tpoly ty) && TypeSet.is_empty (free_univars ty) then
```
Why is this change necessary ?

Removing this change does not seems to affect the testsuite. Would you have an idea of why it is needed and could give a related example ?

(@lpw25 maybe we would need your expertise here)

lpw25 replied:
Otherwise "partial instantiation" can break type checker invariants that require some places to always have `Tpoly` by removing them. I think this is probably an existing bug, but maybe you can't trigger it until this patch.

Octachron replied:
After discussion with @samsa1, I think it is still not possible to reach this case: Tpoly will have the same level as the `Tarrow _` parent node, and thus we will forget the parent node before reaching the Tpoly itself.  However, I agree that this is better to be defensive here.

Nevertheless, we might want to add a case of partial instantiation to the testsuite

```ocaml
let f (x: [< `A of ('a. 'a option) -> unit ]) = match x with `A f -> f None
```
(which gives the empty type ``[< `A of ('a. 'a option) -> unit & 'b option -> 'c ] -> 'c`` to `f`)

samsa1 commented on typing/typecore.ml:
```diff
@@ -3217,9 +3196,20 @@ let loc_rest_of_function
+        (* Polymorphic types will only unify with types that match all of their
+         polymorphic parts, so we need to fully translate the type here
+         unlike in the monomorphic case *)
+        Typetexp.transl_simple_type env ~closed:false arg_sty
```
This change renders the fix in #13984 invalid and makes the related bug much harder to fix.
On another side I'm quite surprised of doing an exact translation here when we are only doing a partial type approximation.

This kind of approximation seems no that necessary. On example we found that needed this is :
```ocaml
let rec f () = g () Fun.id
and g () : ('a. 'a -> 'a) -> unit = fun _ -> ()
```
Which might not be the most common pattern.
Maybe doing a minimal version here with a test saying it could be improved and doing a new PR afterward to improve expressivity

lpw25 replied:
#13984 looks a bit dubious to me. It seems to ignore the possibility of shadowing and assume that using an incorrect type environment can only cause lookup failures. I would expect us to stop pushing types doing approximation whenever we change the type environment.

I think you need this part to support any polymorphic parameter in the type annotation of a recursive function.
```ocaml
let rec foo : ('a. 'a -> 'a) -> () = fun _ -> ()
```
and the error you'd get if you didn't do it would be nonsense.

lpw25 replied:
Or I suppose you are suggesting to abandon the `Tarrow` part if the parameter is polymorphic? That wouldn't break the example I gave there, but would break any such example which actually used the recursion:
```
let rec foo : ('a. 'a -> 'a) -> () = fun f -> foo f 
```

samsa1 commented on typing/typecore.ml:
```diff
@@ -3229,14 +3219,20 @@ let rec approx_type env sty =
+        match sty with
+        | {ptyp_desc=Ptyp_poly _} ->
+          let inferred_ty =
+            Typetexp.transl_simple_type env ~closed:false sty
```
This lines makes a difference between the two following examples :
```ocaml
let rec f () = g (module Map.Make(Int))
and g (m : (module Map.S)) = ();;
```
and
```ocaml
let rec f () = g (module Map.Make(Int))
and g (m : ('a. module Map.S)) = ();;
```
Where the second one is well-typed, while the was rejected by the type-checker.

lpw25 replied:
This is true, but not a new "feature" unfortunately. IIRC this should also type check:
```ocaml
type t = (module Map.S)
let rec f () = g (module Map.Make(Int))
and g (m : t) = ();;
```

Octachron commented on typing/typecore.ml:
```diff
@@ -7584,6 +7588,9 @@ let report_error ~loc env = function
         Style.inline_code l
+  | Optional_poly_param ->
+      Location.errorf ~loc
+        "Optional parameters cannot be polymorphic"
```
I would propose to keep the name of the label, and optionally the names of the bound universal variables.

voodoos replied:
I am now using the label names in error messages (13e9266ba2cb2df269926f92a168b728365649fd)

samsa1 commented on typing/typecore.ml:
```diff
@@ -5857,24 +5857,63 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
+                false
+              with Unify _ -> true
+            in
+            Btype.backtrack snap;
```
Maybe define a function `is_real_poly` in ctype.ml ?

samsa1 commented on typing/typecore.ml:
```diff
@@ -3977,33 +4021,59 @@ and type_expect_
-              | Tfunction_body _ -> newvar ()
-              | Tfunction_cases _ ->
-                newty (Tarrow (Nolabel, newvar (), newvar (), commu_ok)))
+          (* Assert that [ty] is a function, and return its return type. *)
```
The changes below don't seem to directly impact polymorphic parameters as a whole and could/should be send in a separate PR. In any case this does not impact anything in the testsuite and their are no occurences of GADTs in typing-poly/poly_params.ml to test this. Could a related test be added ?

Octachron replied:
I think the refactoring make sense in this context, but I would like the function to be lifted outside of type_expect_ , and an example in the testsuite would be nice too.

Octachron replied:
Concerning the test, here is a small example:
```ocaml
let f: type a. unit -> (a, ('b. 'b -> 'b) -> int) Type.eq ->  a = 
  fun () Equal f -> f 0
```

Octachron replied:
I still think that the test above should be added to the testsuite to test the interaction between syntactic arity mismatch and polymorphic parameters.

voodoos committed:
review: reorder functions in btype

voodoos committed:
review: comment format

voodoos committed:
review: add a Btype.newgenmono utility symetrical to Ctype.newmono

voodoos committed:
review: use label names in error messages

voodoos committed:
review: additional tests

lpw25 commented:
It would be good to include this fix: https://github.com/ocaml-flambda/flambda-backend/pull/4102

Octachron committed:
Fix subtyping error trace

voodoos committed:
Show issue with subtyping trace

voodoos committed:
tests: Promote error messages improvements after rebasing.

voodoos committed:
review: more reuse in parser and reloc in [param_type]

voodoos committed:
review: move [is_really_poly] utility to Ctype.

voodoos committed:
Add more test cases

voodoos committed:
Move definition closer to usage

voodoos committed:
Additional test cases

Octachron commented on typing/btype.mli:
```diff
@@ -77,6 +77,7 @@ val newty2: level:int -> type_desc -> type_expr
 
 val newgenty: type_desc -> type_expr
         (* Create a generic type *)
+val newgenmono: type_expr -> type_expr
```
A comment would be nice::
```ocaml
(** Create a new generic type wrapped inside a [Tpoly(x,[])] *)
```

Octachron commented on testsuite/tests/typing-poly/poly_params.ml:
```diff
@@ -469,18 +469,4 @@ let () = accept_poly_poly_var poly_poly_var
 [%%expect {|
```
I would propose to add the (rejected) dual example:
```ocaml
let f (`B|_) = ()
let h (f:'a. ([> `A ] as 'a) -> unit ) = f `B
```

Octachron replied:
My comment was unclear, the test is applying `h` to `f`:
```ocaml
let error = h f
```

Octachron commented on typing/typeclass.ml:
```diff
@@ -2203,6 +2211,9 @@ let report_error_doc env ppf =
       (Style.as_inline_code Printtyp.type_scheme) sign.csig_self
+  | Polymorphic_class_parameter ->
+      fprintf ppf
+        "Class parameters cannot be polymorphic"
```
There is a missing dot: "Class parameters cannot be polymorphic."

Octachron commented on typing/ctype.mli:
```diff
@@ -283,9 +290,27 @@ val unify_gadt:
+val filter_arrow_mono: Env.t -> type_expr -> arg_label -> filtered_arrow
+        (* A special case of unification. Composition of [filter_arrow]
+           with [filter_mono] on the argument type. Raises
+           [Filter_arrow_mono_failed] instead of [Unify] *)
```
It seems that this function is only used to implement `is_ignore` in `Typecore`. I would propose to inline its definition in `is_ignore` to avoid creating a public exception just for this function.

Octachron commented on typing/ctype.mli:
```diff
@@ -94,6 +94,7 @@ val new_global_var: ?name:string -> unit -> type_expr
            (as type variables ['a] in type constraints). *)
 val newobj: type_expr -> type_expr
 val newconstr: Path.t -> type_expr list -> type_expr
+val newmono : type_expr -> type_expr
```
A documentation comment would be nice.

Octachron commented on typing/typecore.ml:
```diff
@@ -3977,34 +4029,63 @@ and type_expect_
+                  | Not_a_function ->
+                      let tarrow =
+                        (newty
+                          (Tarrow (arg_label, newvar (), newvar (), commu_ok)));
```
The first `newvar` should be a `newmono`, even if it does not really matter since the unifcation will fail on the `Tarrow`.

Octachron commented on typing/typecore.ml:
```diff
@@ -5099,37 +5183,52 @@ and type_binding_op_ident env s =
+  in
+  if not has_poly && not (tpoly_is_mono ty_arg) && !Clflags.principal
+      && get_level ty_arg < Btype.generic_level then begin
+    let really_poly = Ctype.is_really_poly env ty_arg in
```
Now that `Ctype.is_really_poly` is a separate function, one option would to merge the two conditions, to make it clearer that we are just checking conditions for emitting a warning:
```ocaml
  if !Clflags.principal 
    && not has_poly && not (tpoly_is_mono ty_arg)
    && get_level ty_arg < Btype.generic_level
    && Ctype.is_really_poly env ty_arg
  then Location.prerr_warning loc (not_principal "this higher-rank function");
```

Octachron commented on typing/typecore.ml:
```diff
@@ -5786,22 +5894,56 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
+            if really_poly then
+              Location.prerr_warning app_loc
+                (not_principal "applying a higher-rank function here");
+          end;
```
Same as above, this could be written
```ocaml
if !Clflags.principal 
  && get_level ty_arg < Btype.generic_level 
  && Ctype.is_really_poly env ty_arg
then Location.prerr_warning app_loc (not_principal "applying a higher-rank function here");
```

Octachron commented on typing/ctype.mli:
```diff
@@ -283,9 +290,27 @@ val unify_gadt:
+        (* A special case of unification with [l:'a -> 'b]. If
+           [force_poly] is false then the usual invariant that the
+           argument type be a [Tpoly] node is not enforced. Raises
+           [Filter_arrow_failed] instead of [Unify].  *)
```
I would propose to invert the `force_tpoly` condition, and rename it `param_hole`
```ocaml
val filter_arrow: Env.t -> type_expr -> arg_label -> param_hole:bool ->
         filtered_arrow
        (* A special case of unification with [l:'a -> 'b]. If [param_hole] is
           true then ['a] might be initialized with a [Tvar _] hole to be filled
           later by a [Tpoly _].
           Raises [Filter_arrow_failed] instead of [Unify]. *)
```
with the aim to have at the various call sites:
```ocaml
filter_arrow env (instance ty_expected) arg_label ~param_hole:has_poly
```

Octachron commented on typing/ctype.mli:
```diff
@@ -283,9 +290,27 @@ val unify_gadt:
-           [Filter_arrow_failed] instead of [Unify]. *)
+
+type filtered_arrow =
+  { ty_arg : type_expr;
```
Following the nomenclature that we call `param` the types of parameter that ought to be wrapped in a `Tpoly _`, this is a `ty_param` rather than a `ty_arg`.

Octachron commented on typing/typecore.ml:
```diff
@@ -2823,14 +2815,14 @@ type untyped_apply_arg =
       {
         sarg : Parsetree.expression;
-        ty_arg : type_expr;
+        ty_arg_mono : type_expr;
```
I would propose to keepthe field as  `ty_arg`, since this is a not type wrapped in `Tpoly(_, [])`.

Octachron commented on typing/typecore.ml:
```diff
@@ -3717,6 +3743,32 @@ let with_explanation explanation f =
+   [f = fun x_1 ... x_n -> e] is [a_1 -> a_2 -> ... -> a_n -> b], and we're
+   currently typechecking [a_i -> a_{i+1} -> ... -> b] for [i <= n].
+*)
+type split_function_ty =
```
Those comments should be documentation comments.

Octachron commented on typing/typecore.ml:
```diff
@@ -2336,19 +2323,24 @@ let type_pattern_list
-  end;
-  List.iter (fun f -> f()) tps.tps_pattern_force;
-  if is_optional l then unify_pat val_env pat (type_option (newvar ()));
+  let pvs, pat =
```
I would suggest to expand the `pvs` name into `pattern_variables`.

Octachron commented on typing/typecore.ml:
```diff
@@ -5786,22 +5894,56 @@ and type_argument ?explanation ?recarg env sarg ty_expected' ty_expected =
+              let arg =
+                type_argument env sarg ty_arg' ty_arg0'
+              in
+              arg, ty_arg, vars
```
`vars` should be `vars0` here: in `check_univars env name exp ty_expected vars`, `vars` is the set of univars that we instantiated in `exp`.

Octachron reviewed (commented):
After reading the code, I agree that it is correct and implement the proposed specification. There are however few renaming and code reorganization that would improve readability that I recommend to implement before approving fully and merging.

I propose to let the question of let operators for a subsequent PRs.
Similarly, there is probably some more needed to clean up the specification and implementation of `type_approx` but this is a question anterior to this PR.

OlivierNicole committed:
Document function in comment

OlivierNicole committed:
Add test cases

OlivierNicole committed:
Fix typo in error message

OlivierNicole committed:
Rename function and move it to Btype

OlivierNicole committed:
Inline function and exception used only once

OlivierNicole committed:
Document function

OlivierNicole committed:
Enforce Tpoly invariant

OlivierNicole committed:
Clarify conditions

OlivierNicole committed:
Clearer argument label and field names

OlivierNicole committed:
Turn ordinary comment into a documentation comment

OlivierNicole committed:
Clearer variable name

OlivierNicole committed:
Fix wrong set of univars

OlivierNicole commented:
@voodoos is off at the moment, so, with his approval, I have applied the changes that you requested.

OlivierNicole committed:
review: De-duplicate parser rule for let

OlivierNicole committed:
review: Add test with syntactic arity mismatch

OlivierNicole committed:
review: Fix test

OlivierNicole commented:
Please let me know if there are comments to address that I have missed.

Octachron reviewed (approved):
All the review comments have been addressed, the PR is ready to be merged.
I may propose some refactoring later, but this will better handled as subsequent PRs.

The commit history could be simplified, but personally I am inclined to merge the PR with the current history except if @OlivierNicole or @voodoos really wish to clean the history.

voodoos commented:
I have other things than this commit history to clean right now, so if it's good for you it's good for me 👍

Octachron commented:
While rebasing to fix conflicts, I took the liberty to squash the three commits that were purely "diff reduction" commits. I am planning to merge as soon as the CI is green.
